src/core/tsi/ssl_transport_security.cc

```diff
@@ -1410,6 +1410,28 @@ static void ssl_handshaker_destroy(tsi_handshaker* self) {
   gpr_free(impl);
 }
 
+static tsi_result ssl_bytes_remaining(tsi_ssl_handshaker* impl,
+                                      unsigned char** bytes_remaining,
+                                      size_t* bytes_remaining_size) {
+  if (impl == nullptr || bytes_remaining == nullptr ||
+      bytes_remaining_size == nullptr) {
+    return TSI_INVALID_ARGUMENT;
+  }
+  tsi_result result = TSI_OK;
+  size_t counter = 0;
+  size_t bytes_in_ssl_buffer = BIO_pending(SSL_get_rbio(impl->ssl));
+  if (bytes_in_ssl_buffer == 0) return TSI_OK;
+  *bytes_remaining = static_cast<uint8_t*>(gpr_malloc(bytes_in_ssl_buffer));
+  int read_success = 1;
+  while (read_success > 0 && counter < bytes_in_ssl_buffer) {
+    read_success =
+        BIO_read(SSL_get_rbio(impl->ssl), *bytes_remaining + counter, 1);
+    if (read_success == 1) counter += 1;
+  }
+  *bytes_remaining_size = counter;
```

Comment:
Is it expected that  *bytes_remaining_size == bytes_in_ssl_buffer in the end of the call? If so, should you check that?

Reply:
Yes, good call. That's what should occur and I have added a check to this effect. Now, if that is not the case, then we return an error.